src/Native/Runtime/arm64/WriteBarriers.S

```diff
@@ -1,33 +1,368 @@
+// successful updates. 
+
+// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpCheckedLockCmpXchgAVLocation
+// - Function "UnwindWriteBarrierToCaller" assumes no registers where pushed and LR contains the return address
+
+// RhpCheckedLockCmpXchg(Object** dest, Object* value, Object* comparand)
+//
+// Interlocked compare exchange on objectref.
+//
+// On entry:
+//  x0: pointer to objectref
+//  x1: exchange value
+//  x2: comparand
+//
+// On exit:
+//  x0: original value of objectref
+//  x9: trashed
+//  x10: trashed
+//
+    LEAF_ENTRY RhpCheckedLockCmpXchg
+    ALTERNATE_ENTRY  RhpCheckedLockCmpXchgAVLocation
+
+CmpXchgRetry:
+        // Check location value is what we expect.
+        ldaxr   x10, [x0]
+        cmp     x10, x2
+        bne     CmpXchgNoUpdate
+
+        // Current value matches comparand, attempt to update with the new value.
+        stlxr   w9, x1, [x0]
+        cbnz    w9, CmpXchgRetry
+
+        // We have successfully updated the value of the objectref so now we need a GC write barrier.
+        // The following barrier code takes the destination in x0 and the value in x1 so the arguments are
+        // already correctly set up.
+
+        INSERT_CHECKED_WRITE_BARRIER_CORE x0, x1, 9, x0
+
+CmpXchgNoUpdate:
+        // x10 still contains the original value.
+        mov     x0, x10
+        ret     lr
+
+    LEAF_END RhpCheckedLockCmpXchg, _TEXT
+
+// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen within at RhpCheckedXchgAVLocation
+// - Function "UnwindWriteBarrierToCaller" assumes no registers where pushed and LR contains the return address
+ 
+// RhpCheckedXchg(Object** destination, Object* value)
+//
+// Interlocked exchange on objectref.
+//
+// On entry:
+//  x0: pointer to objectref
+//  x1: exchange value
+//
+// On exit:
+//  x0: original value of objectref
+//  x9: trashed
+//  x10: trashed
+//
+    LEAF_ENTRY RhpCheckedXchg, _TEXT
+    ALTERNATE_ENTRY  RhpCheckedXchgAVLocation
+
+ExchangeRetry:
+        // Read the existing memory location.
+        ldaxr   x10,  [x0]
+
+        // Attempt to update with the new value.
+        stlxr   w9, x1, [x0]
+        cbnz    w9, ExchangeRetry
+
+        // We have successfully updated the value of the objectref so now we need a GC write barrier.
+        // The following barrier code takes the destination in x0 and the value in x1 so the arguments are
+        // already correctly set up.
+
+        INSERT_CHECKED_WRITE_BARRIER_CORE x0, x1, 9, x0
+
+        // x10 still contains the original value.
+        mov     x0, x10
+        ret
+
+    LEAF_END RhpCheckedXchg, _TEXT
+
+LEAF_ENTRY RhpAssignRefArm64, _TEXT
```

Comment:
This needs the AV location to handle AVs. 

It may be easiest to just forward the `RhpAssignRef` to here, so that we do not need duplicated set of labels.

Reply:
I am sorry I am not sure if I understand this request correctly.

The RhpAssignRefAVLocation is set as an alternative name for RhpAssignRef like in the asm version. I added the RhpAssignRefARm64 variant as RyuJIT is using a custom calling convention by using X14 and X15 instead of X0 and X1.

Reply:
The code that converts segfaults signals into NullReferenceExceptions looks at IP of segfault to decide whether it is safe to do so. We do not want to be converting segfaults in random unmanaged code into NullReferenceExceptions.

The IP has to be either in the managed code, or in one of the special locations in hand-written assembly code. `RhpAssignRef...` is one of those special locations where we want to convert segfaults into NullReferenceExceptions.

The problem is that `RhpAssign{Checked}RefArm64` as written is not marked as one these special locations. There are two options how it can be fixed:

Option 1:
- Add ALTERNATE_ENTRY for `RhpAssign{Checked}RefArm64`
- Add handling of these entries to `InWriteBarrierHelper` in EHHelpers.cpp under ifdef

Option 2:
- Change Rhp{Checked}AssignRef to be:
```
LEAF_ENTRY Rhp{Checked}AssignRef
    mov     x14, x0                     ; x14 = dst
    mov     x15, x1                     ; x15 = val
    b       Rhp{Checked}AssignRefArm64
```
- Move `Rhp{Checked}AssignRefAVLocation` to Rhp{Checked}AssignRefArm64, no changes in EHHelpers.cpp required

Pick the one you like better, either one is fine with me.

Reply:
OK, I think I understand what needs to be done there.